rmw_dds_common/src/time_utils.cpp

```diff
@@ -24,21 +24,40 @@ rmw_dds_common::clamp_rmw_time_to_dds_time(const rmw_time_t & time)
 {
   rmw_time_t t = time;
 
-  // Let's try to keep the seconds value representable with 32-bits
-  // by moving the excess seconds over to the nanoseconds field.
-  if (t.sec > INT_MAX) {
-    uint64_t diff = t.sec - INT_MAX;
-    t.sec -= diff;
-    t.nsec += diff * (1000LL * 1000LL * 1000LL);
+  // Normalize rmw_time_t::nsec to be < 1s, so that it may be safely converted
+  // to a DDS Duration or Time (see DDS v1.4 section 2.3.2).
+  // If the total length in seconds cannot be represented by DDS (which is
+  // limited to INT_MAX seconds + (10^9 - 1) nanoseconds) we must truncate
+  // the seconds component at INT_MAX, while also normalizing nanosec to < 1s.
+  static const uint64_t sec_to_ns = 1000000000ULL;
+  uint64_t ns_sec_adjust = t.nsec / sec_to_ns;
+  bool overflow_nsec = false,
+    overflow_sec = false;
```

Comment:
@asorbini nit^N:
```suggestion
  bool overflow_nsec = false;
  bool overflow_sec = false;
```
is more verbose but easier to read IMHO.

Reply:
Applied in [1fb9751](https://github.com/ros2/rmw_dds_common/pull/48/commits/1fb97517ff4ed002e510664b6906deae39c4322f).